debian/cvise.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH CVISE "1" "November 2024" "cvise 2.11.0" "User Commands"
.SH NAME
cvise \- super-parallel Python port of the C-Reduce project
.SH DESCRIPTION
usage: cvise [\-h] [\-\-n N] [\-\-tidy] [\-\-shaddap] [\-\-die\-on\-pass\-bug] [\-\-sllooww]
.IP
[\-\-also\-interesting EXIT_CODE] [\-\-debug]
[\-\-log\-level {INFO,DEBUG,WARNING,ERROR}] [\-\-log\-file LOG_FILE]
[\-\-no\-give\-up] [\-\-print\-diff] [\-\-save\-temps]
[\-\-skip\-initial\-passes] [\-\-skip\-interestingness\-test\-check]
[\-\-remove\-pass REMOVE_PASS] [\-\-start\-with\-pass START_WITH_PASS]
[\-\-no\-timing] [\-\-timestamp] [\-\-timeout [TIMEOUT]] [\-\-no\-cache]
[\-\-skip\-key\-off] [\-\-max\-improvement BYTES]
[\-\-pass\-group {all,binary,opencl\-120,delta} | \fB\-\-pass\-group\-file\fR PASS_GROUP_FILE]
[\-\-clang\-delta\-std {c++98,c++11,c++14,c++17,c++20,c++2b}]
[\-\-clang\-delta\-preserve\-routine CLANG_DELTA_PRESERVE_ROUTINE]
[\-\-not\-c] [\-\-renaming] [\-\-list\-passes] [\-\-version]
[\-\-commands COMMANDS] [\-\-shell SHELL] [\-\-to\-utf8]
[\-\-skip\-after\-n\-transforms SKIP_AFTER_N_TRANSFORMS]
[\-\-stopping\-threshold STOPPING_THRESHOLD]
[INTERESTINGNESS_TEST] TEST_CASE [TEST_CASE ...]
.PP
C\-Vise
.SS "positional arguments:"
.TP
INTERESTINGNESS_TEST
Executable to check interestingness of test cases
.TP
TEST_CASE
Test cases
.SS "options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-\-n\fR N, \fB\-n\fR N
Number of cores to use; C\-Vise tries to automatically
pick a good setting but its choice may be too low or
high for your situation
.TP
\fB\-\-tidy\fR
Do not make a backup copy of each file to reduce as
file.orig
.TP
\fB\-\-shaddap\fR
Suppress output about non\-fatal internal errors
.TP
\fB\-\-die\-on\-pass\-bug\fR
Terminate C\-Vise if a pass encounters an otherwise
non\-fatal problem
.TP
\fB\-\-sllooww\fR
Try harder to reduce, but perhaps take a long time to
do so
.TP
\fB\-\-also\-interesting\fR EXIT_CODE
A process exit code (somewhere in the range 64\-113
would be usual) that, when returned by the
interestingness test, will cause C\-Vise to save a copy
of the variant
.TP
\fB\-\-debug\fR
Print debug information (alias for \fB\-\-log\-level\fR=\fI\,DEBUG\/\fR)
.TP
\fB\-\-log\-level\fR {INFO,DEBUG,WARNING,ERROR}
Define the verbosity of the logged events
.TP
\fB\-\-log\-file\fR LOG_FILE
Log events into LOG_FILE instead of stderr. New events
are appended to the end of the file
.TP
\fB\-\-no\-give\-up\fR
Don't give up on a pass that hasn't made progress for
50000 iterations
.TP
\fB\-\-print\-diff\fR
Show changes made by transformations, for debugging
.TP
\fB\-\-save\-temps\fR
Don't delete \fI\,/tmp/cvise\-xxxxxx\/\fP directories on
termination
.TP
\fB\-\-skip\-initial\-passes\fR
Skip initial passes (useful if input is already
partially reduced)
.TP
\fB\-\-skip\-interestingness\-test\-check\fR, \fB\-s\fR
Skip initial interestingness test check
.TP
\fB\-\-remove\-pass\fR REMOVE_PASS
Remove all instances of the specified passes from the
schedule (comma\-separated)
.TP
\fB\-\-start\-with\-pass\fR START_WITH_PASS
Start with the specified pass
.TP
\fB\-\-no\-timing\fR
Do not print timestamps about reduction progress
.TP
\fB\-\-timestamp\fR
Print timestamps instead of relative time from a
reduction start
.TP
\fB\-\-timeout\fR [TIMEOUT]
Interestingness test timeout in seconds
.TP
\fB\-\-no\-cache\fR
Don't cache behavior of passes
.TP
\fB\-\-skip\-key\-off\fR
Disable skipping the rest of the current pass when 's'
is pressed
.TP
\fB\-\-max\-improvement\fR BYTES
Largest improvement in file size from a single
transformation that C\-Vise should accept (useful only
to slow C\-Vise down)
.TP
\fB\-\-pass\-group\fR {all,binary,opencl\-120,delta}
Set of passes used during the reduction
.TP
\fB\-\-pass\-group\-file\fR PASS_GROUP_FILE
JSON file defining a custom pass group
.TP
\fB\-\-clang\-delta\-std\fR {c++98,c++11,c++14,c++17,c++20,c++2b}
Specify clang_delta C++ standard, it can rapidly speed
up all clang_delta passes
.TP
\fB\-\-clang\-delta\-preserve\-routine\fR CLANG_DELTA_PRESERVE_ROUTINE
Preserve the given function in replace\-function\-defwith\-decl clang delta pass
.TP
\fB\-\-not\-c\fR
Don't run passes that are specific to C and C++, use
this mode for reducing other languages
.TP
\fB\-\-renaming\fR
Enable all renaming passes (that are disabled by
default)
.TP
\fB\-\-list\-passes\fR
Print all available passes and exit
.TP
\fB\-\-version\fR
show program's version number and exit
.TP
\fB\-\-commands\fR COMMANDS, \fB\-c\fR COMMANDS
Use shell commands instead of an interestingness test
case
.TP
\fB\-\-shell\fR SHELL
Use selected shell for the \fB\-\-commands\fR option
.TP
\fB\-\-to\-utf8\fR
Convert any non\-UTF\-8 encoded input file to UTF\-8
.TP
\fB\-\-skip\-after\-n\-transforms\fR SKIP_AFTER_N_TRANSFORMS
Skip each pass after N successful transformations
.TP
\fB\-\-stopping\-threshold\fR STOPPING_THRESHOLD
CVise will stop reducing a test case once it has
reduced by this fraction of its original size. Between
0.0 and 1.0.
.SS "available shortcuts:"
.IP
S \- skip execution of the current pass
D \- toggle \fB\-\-print\-diff\fR option
.PP
For bug reporting instructions, please use:
https://github.com/marxin/cvise/


debian/changelog
cvise (2.11.20250409) plucky; urgency=medium

  * Dynamically adapt timeouts.
  * Improve child process termination.
  * Fix crash on UTF decoding error.
  * Improve orchestration performance.

 -- Maksim Ivanov <emaxx@google.com>  Wed, 09 Apr 2025 12:00:00 +0100

cvise (2.11.20250403) plucky; urgency=medium

  * Optimizations and multi-file support.

 -- Maksim Ivanov <emaxx@google.com>  Fri, 03 Apr 2025 12:00:00 +0100

cvise (2.11.0-1ubuntu1) plucky; urgency=medium

  * Build using LLVM 20, patch taken from the trunk.

 -- Matthias Klose <doko@ubuntu.com>  Fri, 21 Feb 2025 12:05:34 +0100

cvise (2.11.0-1) unstable; urgency=medium

  * New upstream version.
  * Bump standards version.

 -- Matthias Klose <doko@debian.org>  Tue, 12 Nov 2024 17:44:19 +0100

cvise (2.10.0+20240909-1) unstable; urgency=medium

  * New upstream snapshot.
  * Build using LLVM 19.

 -- Matthias Klose <doko@debian.org>  Mon, 09 Sep 2024 17:25:52 +0200

cvise (2.10.0-2) unstable; urgency=medium

  * Drop build dependency on python-pytest-flake8 (Olivier Gayot).

 -- Matthias Klose <doko@debian.org>  Mon, 15 Apr 2024 15:47:33 +0200

cvise (2.10.0-1) unstable; urgency=medium

  * New upstream version.

 -- Matthias Klose <doko@debian.org>  Wed, 13 Mar 2024 21:28:24 +0100

cvise (2.9.0-4) unstable; urgency=medium

  * Build using LLVM 18.

 -- Matthias Klose <doko@debian.org>  Tue, 12 Mar 2024 16:42:54 +0100

cvise (2.9.0-2) unstable; urgency=medium

  * Build using LLVM 17.0.6.

 -- Matthias Klose <doko@debian.org>  Sun, 10 Dec 2023 19:29:46 +0100

cvise (2.9.0-1) unstable; urgency=medium

  * New upstream version.
  * Build using LLVM 17.

 -- Matthias Klose <doko@debian.org>  Thu, 16 Nov 2023 14:08:12 +0100

cvise (2.8.0-2) unstable; urgency=medium

  * Build using LLVM 16.
  * Bump standards version.

 -- Matthias Klose <doko@debian.org>  Thu, 24 Aug 2023 19:10:09 +0200

cvise (2.8.0-1) unstable; urgency=medium

  * New upstream version.

 -- Matthias Klose <doko@debian.org>  Tue, 06 Jun 2023 17:23:25 +0200

cvise (2.7.0-2) unstable; urgency=medium

  * Don't call dh_dwz on mipsel (runs out of memory).

 -- Matthias Klose <doko@debian.org>  Sat, 31 Dec 2022 10:52:35 +0100

cvise (2.7.0-1) unstable; urgency=medium

  * New upstream version.
  * Drop the excessive llvm build dependencies.

 -- Matthias Klose <doko@debian.org>  Mon, 26 Dec 2022 11:39:03 +0100

cvise (2.6.0-1) unstable; urgency=medium

  * New upstream version.
  * Build using LLVM 15.
  * Build-depend on clang-tools-15, clang-tidy-15, clangd-15. These are not
    used in anyway, but LLVM's broken cmake support requires these.

 -- Matthias Klose <doko@debian.org>  Sun, 23 Oct 2022 20:33:43 +0200

cvise (2.5.0-1) unstable; urgency=medium

  * New upstream version.

 -- Matthias Klose <doko@debian.org>  Wed, 14 Sep 2022 16:37:13 +0200

cvise (2.4.0-3) unstable; urgency=medium

  * Build using LLVM 14.

 -- Matthias Klose <doko@debian.org>  Fri, 18 Feb 2022 00:59:37 +0100

cvise (2.4.0-2.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Remove stale LLVM 9 build dependencies on armel/armhf.
    (Closes: #1000892)
  * Recommend colordiff. (Closes: #1001375)

 -- Adrian Bunk <bunk@debian.org>  Wed, 29 Dec 2021 22:41:49 +0200

cvise (2.4.0-2) unstable; urgency=medium

  * (Build-)depend on python3-chardet.

 -- Matthias Klose <doko@debian.org>  Tue, 02 Nov 2021 16:15:02 +0200

cvise (2.4.0-1) unstable; urgency=medium

  * New upstream version.
  * Build using LLVM 13.
  * Bump standards version.

 -- Matthias Klose <doko@debian.or>  Tue, 02 Nov 2021 14:43:50 +0200

cvise (2.3.0-1) unstable; urgency=medium

  * New upstream version.
  * Can be built using LLVM 12, which also fixes #979495.

 -- Matthias Klose <doko@debian.org>  Wed, 17 Mar 2021 12:47:09 +0100

cvise (2.2.0-2) unstable; urgency=medium

  * Fix issue #47, remove temporary files in working directory.

 -- Matthias Klose <doko@debian.org>  Mon, 08 Mar 2021 16:41:40 +0100

cvise (2.2.0-1) unstable; urgency=medium

  * New upstream version.

 -- Matthias Klose <doko@debian.org>  Fri, 05 Mar 2021 18:25:42 +0100

cvise (2.1.0-1) unstable; urgency=medium

  * New upstream version.

 -- Matthias Klose <doko@debian.org>  Tue, 12 Jan 2021 19:21:35 +0100

cvise (2.0.0-4) unstable; urgency=medium

  * Update to the branch 20210108.
  * Build using LLVM 9 on armel, armhf. See #979495.
  * Bump debhelper version.

 -- Matthias Klose <doko@debian.org>  Fri, 08 Jan 2021 14:02:42 +0100

cvise (2.0.0-3) unstable; urgency=medium

  * Update to the branch 20201126.
    - Handle using declarations declared with macros.
  * Bump standards version.

 -- Matthias Klose <doko@debian.org>  Sun, 13 Dec 2020 12:37:58 +0100

cvise (2.0.0-2) unstable; urgency=medium

  * New upstream version.
  * Updates up to 7c303ae036e18be10ab4014724b9e14d4a1245d6.

 -- Matthias Klose <doko@debian.org>  Tue, 17 Nov 2020 14:09:28 +0100

cvise (1.9.0-1) unstable; urgency=medium

  * New upstream version.
  * Updates up to a000f9eefce217576ac3095a443fcd9bb7b95953.

 -- Matthias Klose <doko@debian.org>  Tue, 10 Nov 2020 12:03:35 +0100

cvise (1.8.0-1) unstable; urgency=medium

  * New upstream version.
  * Build using LLVM 11.

 -- Matthias Klose <doko@debian.org>  Thu, 22 Oct 2020 13:17:58 +0200

cvise (1.7.0-1) unstable; urgency=medium

  * New upstream version.

 -- Matthias Klose <doko@debian.org>  Tue, 06 Oct 2020 14:00:36 +0200

cvise (1.6.0-3) unstable; urgency=medium

  * Pull fixes from the VCS 20200929:
    - Fix psutil core count detection code.

 -- Matthias Klose <doko@debian.org>  Tue, 29 Sep 2020 16:37:45 +0200

cvise (1.6.0-2) unstable; urgency=medium

  * Pull fixes from the VCS 20200917:
    - Fix infinite loop in IfPass.

 -- Matthias Klose <doko@debian.org>  Thu, 17 Sep 2020 13:42:14 +0200

cvise (1.6.0-1) unstable; urgency=medium

  * New upstream version.

 -- Matthias Klose <doko@debian.org>  Sun, 02 Aug 2020 13:13:22 +0200

cvise (1.5.0-1) unstable; urgency=medium

  * New upstream version.


cvise/cvise.py
import json
import logging
import os
import sys

from cvise.passes.abstract import AbstractPass
from cvise.passes.balanced import BalancedPass
from cvise.passes.blank import BlankPass
from cvise.passes.clang import ClangPass
from cvise.passes.clangbinarysearch import ClangBinarySearchPass
from cvise.passes.clang_unused_code import ClangUnusedCodePass
from cvise.passes.clex import ClexPass
from cvise.passes.comments import CommentsPass
from cvise.passes.delete_file import DeleteFilePass
from cvise.passes.gcdabinary import GCDABinaryPass
from cvise.passes.generic import GenericPass
from cvise.passes.ifs import IfPass
from cvise.passes.includeincludes import IncludeIncludesPass
from cvise.passes.includes import IncludesPass
from cvise.passes.indent import IndentPass
from cvise.passes.ints import IntsPass
from cvise.passes.line_markers import LineMarkersPass
from cvise.passes.lines import LinesPass
from cvise.passes.peep import PeepPass
from cvise.passes.special import SpecialPass
from cvise.passes.ternary import TernaryPass
from cvise.passes.tree_sitter import TreeSitterPass
from cvise.passes.unifdef import UnIfDefPass
from cvise.utils.error import CViseError, PassOptionError


class CVise:
    class Info:
        PACKAGE_BUGREPORT = '@cvise_PACKAGE_BUGREPORT@'
        PACKAGE_NAME = '@cvise_PACKAGE_NAME@'
        PACKAGE_STRING = '@cvise_PACKAGE_STRING@'
        PACKAGE_URL = '@cvise_PACKAGE_URL@'
        PACKAGE_VERSION = '@cvise_PACKAGE_VERSION@'

        VERSION = '@cvise_VERSION@'
        GIT_VERSION = '@GIT_HASH@'
        LLVM_VERSION = '@LLVM_VERSION@'

    pass_name_mapping = {
        'balanced': BalancedPass,
        'blank': BlankPass,
        'clang': ClangPass,
        'clangbinarysearch': ClangBinarySearchPass,
        'clang_unused_code': ClangUnusedCodePass,
        'clex': ClexPass,
        'comments': CommentsPass,
        'delete-file': DeleteFilePass,
        'gcda-binary': GCDABinaryPass,
        'generic': GenericPass,
        'ifs': IfPass,
        'includeincludes': IncludeIncludesPass,
        'includes': IncludesPass,
        'indent': IndentPass,
        'ints': IntsPass,
        'line_markers': LineMarkersPass,
        'lines': LinesPass,
        'peep': PeepPass,
        'special': SpecialPass,
        'ternary': TernaryPass,
        'tree_sitter': TreeSitterPass,
        'unifdef': UnIfDefPass,
    }

    def __init__(self, test_manager, skip_interestingness_test_check):
        self.test_manager = test_manager
        self.skip_interestingness_test_check = skip_interestingness_test_check
        self.tidy = False

    @classmethod
    def load_pass_group_file(cls, path):
        with open(path) as pass_group_file:
            try:
                pass_group_dict = json.load(pass_group_file)
            except json.JSONDecodeError:
                raise CViseError('Not valid JSON.') from None

        return pass_group_dict

    @classmethod
    def parse_pass_group_dict(
        cls,
        pass_group_dict,
        pass_options,
        external_programs,
        remove_pass,
        clang_delta_std,
        clang_delta_preserve_routine,
        not_c,
        renaming,
    ):
        pass_group = {}
        removed_passes = set(remove_pass.split(',')) if remove_pass else set()

        def parse_options(options):
            valid_options = set()

            for opt in options:
                try:
                    valid_options.add(AbstractPass.Option(opt))
                except ValueError:
                    raise PassOptionError(opt) from None

            return valid_options

        def include_pass(pass_dict, options):
            return (('include' not in pass_dict) or bool(parse_options(pass_dict['include']) & options)) and (
                ('exclude' not in pass_dict) or not bool(parse_options(pass_dict['exclude']) & options)
            )

        STANDARD_CATEGORIES = ('first', 'main', 'last')
        PARALLEL_CATEGORIES = ('parallel::initial', 'parallel::main', 'parallel::last')
        for category in STANDARD_CATEGORIES + PARALLEL_CATEGORIES:
            if category not in pass_group_dict:
                if category not in PARALLEL_CATEGORIES:
                    raise CViseError(f'Missing category {category}')
                continue

            pass_group[category] = []

            for pass_dict in pass_group_dict[category]:
                if not include_pass(pass_dict, pass_options):
                    continue

                if 'pass' not in pass_dict:
                    raise CViseError(f'Invalid pass in category {category}')

                try:
                    pass_class = cls.pass_name_mapping[pass_dict['pass']]
                except KeyError:
                    raise CViseError('Unkown pass {}'.format(pass_dict['pass'])) from None

                pass_instance = pass_class(pass_dict.get('arg'), external_programs)
                pass_instance.max_transforms = None
                if str(pass_instance) in removed_passes:
                    continue

                if 'strategy' in pass_dict:
                    pass_instance.strategy = pass_dict['strategy']
                pass_instance.jobs = pass_dict.get('jobs', 1)

                if not_c and 'c' in pass_dict and pass_dict['c']:
                    continue
                elif not renaming and 'renaming' in pass_dict and pass_dict['renaming']:
                    continue

                try:
                    pass_instance.max_transforms = int(pass_dict['max-transforms'])
                except KeyError:
                    pass

                pass_instance.user_clang_delta_std = clang_delta_std
                pass_instance.clang_delta_preserve_routine = clang_delta_preserve_routine
                pass_group[category].append(pass_instance)

        return pass_group

    def reduce(self, pass_group, skip_initial):
        self._check_prerequisites(pass_group)
        if not self.skip_interestingness_test_check:
            self.test_manager.check_sanity(True)

        logging.info(f'===< {os.getpid()} >===')
        logging.info(
            'running {} interestingness test{} in parallel'.format(
                self.test_manager.parallel_tests,
                '' if self.test_manager.parallel_tests == 1 else 's',
            )
        )

        if not self.tidy:
            self.test_manager.backup_test_cases()

        if pass_group.get('parallel::initial') and not skip_initial:
            logging.info('PARALLEL INITIAL PASSES')
            self._run_main_passes(pass_group['parallel::initial'], True)
        if pass_group.get('parallel::main'):
            logging.info('PARALLEL MAIN PASSES')
            self._run_main_passes(pass_group['parallel::main'], True)
        if pass_group.get('parallel::last'):
            logging.info('PARALLEL CLEANUP PASSES')
            self._run_main_passes(pass_group['parallel::last'], True)

        if all(not t.is_dir() for t in self.test_manager.test_cases):
            if not skip_initial:
                logging.info('INITIAL PASSES')
                self._run_additional_passes(pass_group['first'])

            logging.info('MAIN PASSES')
            self._run_main_passes(pass_group['main'])

            logging.info('CLEANUP PASSES')
            self._run_additional_passes(pass_group['last'])

        logging.info('===================== done ====================')
        return True

    @staticmethod
    def _check_prerequisites(pass_group):
        for category in pass_group:
            for p in pass_group[category]:
                if not p.check_prerequisites():
                    logging.error(f'Prereqs not found for pass {p}')

    def _run_additional_passes(self, passes):
        for p in passes:
            if not p.check_prerequisites():
                logging.error(f'Skipping {p}')
            else:
                self.test_manager.run_pass(p)

    def _run_main_passes(self, passes, parallel=False):
        while True:
            total_file_size = self.test_manager.total_file_size
            total_file_count = self.test_manager.total_file_count

            met_stopping_threshold = False
            if parallel:
                self.test_manager.run_concurrent_passes(passes)
            else:
                for p in passes:
                    # Exit early if we're already reduced enough
                    improvement = (
                        self.test_manager.orig_total_file_size - total_file_size
                    ) / self.test_manager.orig_total_file_size
                    logging.debug(
                        f'Termination check: stopping threshold is {self.test_manager.stopping_threshold}; current improvement is {improvement:.1f}'
                    )
                    if improvement >= self.test_manager.stopping_threshold:
                        met_stopping_threshold = True
                        break
                    if not p.check_prerequisites():
                        logging.error(f'Skipping pass {p}')
                    else:
                        self.test_manager.run_pass(p)

            logging.info(f'Termination check: size was {total_file_size} count {total_file_count}; now size {self.test_manager.total_file_size} count {self.test_manager.total_file_count}')

            worse = self.test_manager.total_file_size >= total_file_size and self.test_manager.total_file_count >= total_file_count
            if worse or met_stopping_threshold:
                break
